File: backend/songs/structure.go
package songs

import "fmt"

const Chorus = 0

type SongText struct {
	Tonality Tonality `json:"tonality"`
	Stances  []Stance `json:"stances"`
}

type StanceType int

type Stance struct {
	Text []Word     `json:"verses"`
	Type StanceType `json:"type"`
}

type Word []Syllable

type Syllable struct {
	Text  string `json:"syllable"`
	Chord *Chord `json:"chord"`
}

type Chord struct {
	Root    Note   `json:"root"`
	Base    *Note  `json:"base"`
	Details string `json:"details"`
}

// Chord String
func (c Chord) String() string {
	if c.Base == nil {
		return fmt.Sprintf("%v%s", c.Root, c.Details)
	}
	return fmt.Sprintf("%v%s/%v", c.Root, c.Details, c.Base)
}

// Transposition chain
func (s *SongText) TransposeAllStances(interval Interval) {
	for i := 0; i < len(s.Stances); i++ {
		s.Stances[i].transposeAllWords(interval)
	}
}

func (s *Stance) transposeAllWords(interval Interval) {
	for i := 0; i < len(s.Text); i++ {
		s.Text[i].transposeAllChords(interval)
	}
}

func (w *Word) transposeAllChords(interval Interval) {
	for i := 0; i < len(*w); i++ {
		(*w)[i].transposeChord(interval)
	}
}

func (s *Syllable) transposeChord(i Interval) {
	s.Chord.Transpose(i)
}

func (c *Chord) Transpose(i Interval) {
	if c.Base == nil {
		c.Base.Transpose(i)
	}
	c.Root.Transpose(i)
}

File: frontend/src/SongEditor/song_text/structure.ts
import { Note, Tonality, Interval, NoteName, newNote } from "./transposer";

export type SongText = {
  tonality: Tonality;
  stances: SongStance[];
};

export function transposeSongText(st: SongText, i: Interval): SongText {
  const n = st.tonality.note.transpose(st.tonality.note, i);
  st.tonality.note = n;
  st.stances = st.stances.map((s) => transposeStance(s, i));
  return st;
}

export type Chorus = 0;
export type StanceType = Chorus | number;
export type SongStance = {
  type: StanceType;
  text: Verse[];
};

export function emptySong(): SongText {
  return {
    tonality: {
      details: "Maior",
      note: newNote(NoteName.C, 0),
    },
    stances: [
      {
        type: 0,
        text: [[[{ text: "", chord: makeChord(newNote(NoteName.C, 0)) }]]],
      },
    ],
  };
}

export type Verse = Word[];

function transposeStance(s: SongStance, i: Interval): SongStance {
  s.text = s.text.map((v) => v.map((w) => transposeWords(w, i)));
  return s;
}

export type Word = Syllable[];

function transposeWords(w: Word, i: Interval): Word {
  return w.map((s) => {
    s.chord = s.chord ? s.chord.transpose(s.chord, i) : undefined;
    return s;
  });
}

export type Syllable = {
  text: string;
  chord?: Chord;
};

export function emptySyllable() {
  return {
    text: "",
  };
}

export interface Chord {
  root: Note;
  base?: Note;
  details?: string;
  toString(c: Chord): string;
  transpose(c: Chord, i: Interval): Chord;
}

export default function makeChord(
  root: Note,
  base?: Note,
  details?: string,
): Chord {
  return {
    root,
    base,
    details,
    toString: chordToString,
    transpose: transposeChord,
  };
}

function chordToString(c: Chord) {
  const { root, base, details } = c;
  return (
    root.toString(root) +
    (base ? "/" + base.toString(base) : "") +
    (details ? details : "")
  );
}

function transposeChord(c: Chord, i: Interval): Chord {
  let { root, base } = c;
  root = root.transpose(root, i);
  if (base) {
    base = base.transpose(base, i);
  }
  return {
    root,
    base,
    details: c.details,
    toString: c.toString,
    transpose: c.transpose,
  };
}
